client/src/services/memberService.js

```diff
@@ -0,0 +1,31 @@
+export default class MemberService {
+	static ENDPOINT = "/api/users";
+
+	constructor(request = fetch) {
+		this.fetch = request;
+	}
+
+	async getUserById(id) {
+		let result = {
+			data: undefined,
+			error: undefined,
+		};
```

Comment:
This is a pattern that I personally useful, the one that Tanstack Query uses to make restful calls,  it's not the complete implementation here, but the idea is to wrap the promise and gives feedback on its state to the component using it. the return object is sth like this { data, isSuccess, isError, error, isLoading ,...}, it gives the component the possibility to handle gracefully the state of the external call and eventual errors, I did not implement it due to lack of time.

Reply:
You often see this kind of thing with hooks because they **cannot** throw an error without breaking the render cycle. But although our services _are_ provided via a hook, that's only needed to inject access to React Router (so that the user can be removed from an auth-only route if they're not authenticated, see `client/src/authContext.js` and `client/src/services/index.js`); they don't have any state of their own to keep tracking of loading, etc. This keeps the services themselves pure and easy to understand and test without any React context, and it's up to the page(s) using them to determine their loading states potentially according to multiple requests across multiple services, in this case e.g.:

```javascript
useEffect(() => {
  setLoading(true);
  resourceService
    .getDraftById(draftId)
    .then(({ recommender }) => memberService.getUserById(recommender))
    .then(setUserDetails)
    .catch((err) => /* ??? */)
    .finally(() => setLoading(false));
}, [draftId, memberService, resourceService]);
```

Reply:
thanks for the explanation, it's interesting but doesn't  it lead to huger components and repetition of code  on the component level, when multiple component use the same service? we are using React why not hook to react lifecycle in a centralise way? I mean using custom hooks for services, why do we need to be able to "test without any react context"?

Reply:
It's all trade-offs. It can lead to some repetition, but if multiple components _are_ using the same data we can decide _at that point_ whether it makes sense to have a higher-level hook, or use a context, or something else. This way we get to define our own abstractions for what these transport-level actions mean (e.g. _"PATCH the `draft` property to `false`"_ -> _"publish the resource"_).

Leaning more into using React (or any other third party library/framework) for _everything_ couples you more to that specific framework, which [can sometimes be a problem](https://tanzu.vmware.com/content/blog/react-license-woes). If you write as much code as possible that's independent of _any_ framework, it can easily be used and reused elsewhere, and even maintained by people who don't know anything about that framework. (Plus remember the purpose of this specific app; there's only so many new things we want applicants to have to spend time learning.)

Also at the high level it's about following the patterns you see in the project you're working on, rather than unilaterally changing them - live in the house a little while, before you start knocking through the walls.